src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            addHead(value);
+        } else {
+            addTail(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        } else {
+            addInside(index,value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkValidIndex(index);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkValidIndex(index);
+        T replacedValue = findNodeByIndex(index).item;
+        findNodeByIndex(index).item = value;
+        return replacedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkValidIndex(index);
+        Node<T> current = findNodeByIndex(index);
+        T removedValue = current.item;
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else {
+            unlink(index, current);
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.item == object
+                    || currentNode.item != null && currentNode.item.equals(object)) {
+                remove(i);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.next = next;
+            this.item = item;
+            this.prev = prev;
+        }
+    }
+
+    private void addHead(T value) {
+        head = new Node<>(null, value, null);
+        tail = head;
+        size++;
+    }
+
+    private void addTail(T value) {
```

Comment:
let's remove these methods and move their logic directly to add method, you can see that there are duplications that can be eliminated